Update seed data

stephenrichards opened the pull request.

misaka commented on db/schema.rb:
```diff
@@ -57,6 +57,18 @@
+    t.integer  "item_id",        null: false
+    t.string   "event",          null: false
+    t.string   "whodunnit"
+    t.text     "object"
```
Is it just me or has this column, and `object_changes` below, lost the parameter `limit: TEXT_BYTES` that's in the migration?

misaka commented on db/seeds.rb:
```diff
@@ -1,4 +1,4 @@
-# This file should contain all the record creation needed to seed the database with its default values.
+  # This file should contain all the record creation needed to seed the database with its default values.
```
Might be worth removing this from the PR just to cut down on noise. ;)

misaka commented on app/models/correspondence.rb:
```diff
@@ -37,4 +42,50 @@ def drafter_assigned?
+    history_events.push initial_submission_event
+    history_events += state_change_events
+    history_events += un_paper_trailed_current_state unless un_paper_trailed_current_state.nil?
+    history_events
```
Style-wise, I would be tempted to make this look more functional:

```ruby
    @versions = PaperTrail::Version.where(
      item_type: 'Correspondence',
      item_id:   id
    ).order('ID ASC')

    [
      initial_submission_event,
      *state_change_events,
      *un_paper_trailed_current_state
    ]
```
It just means less intermediary variables, etc. By the same token I'd be tempted to call this methods `history_events`, since it's really just returning a list of events ... it's not acting as an agent, you can't give it options, etc.

misaka commented on app/models/correspondence.rb:
```diff
@@ -37,4 +42,50 @@ def drafter_assigned?
+      end
+      prior_state = this_version.state
+    end
+    change_events
```
Similar to my thoughts on the style for `historian`, this could be rewritten in a more functional style:

```ruby
  def state_change_events
    prior_state = 'submitted'
    change_events = @versions.find_all do |version|
      version.state != prior_state ? prior_state = version.state : false
    end
    change_events.map do |version|
      email = User.find_by_id(version.whodunnit).email
      NotedEvent.new(email, version.created_at, version.state)
    end
  end
```

misaka commented on app/models/correspondence.rb:
```diff
@@ -37,4 +42,50 @@ def drafter_assigned?
+        state
+      )
+    end
+    last_change_is_state_change
```
Wouldn't be complete if this wasn't also made more functional style too:

```ruby
  def un_paper_trailed_current_state
    # Note that paper_trail records the data prior to a change in 'version'.
    # - so if the last change to a record is to state, that would not show in the version table.
    if @versions.last.state != state
      [
        NotedEvent.new(
          User.find_by_id(user_id).email,
          created_at,
          state
        )
      ]
    else
      []
    end
  end
```

misaka replied:
Actually, rubocop whinges about this method being too long. This would look better:

```ruby
  def un_paper_trailed_current_state
    # Note that paper_trail records the data prior to a change in 'version'.
    # - so if the last change to a record is to state, that would not show in the version table.
    if @versions.last.state != state
      email = User.find_by_id(user_id).email
      [NotedEvent.new(email, created_at, state)]
    else
      []
    end
  end
```

misaka commented on spec/features/correspondence_assignation_spec.rb:
```diff
@@ -12,7 +12,9 @@
+    expect(page.case_history_list.count).to eq 1
     click_button 'Assign'
     expect(page).to have_content("Correspondence assigned to #{User.first.email}")
+    expect(page.case_history_list.count).to eq 2
```
I'd expect to see tests for the actual content of the case history here too, not just counting how many show up.

aliuk2012 replied:
Isn't that what line 17 is doing? I think line 15  (i.e before you assign) should test the content of the case history so it would fail if say we had a new case that didn't have a "create" event but was just assigned directly to a user.

misaka commented:
You've added a good bit of code into the model to do stuff like generating history events, etc. I think it'd be worth adding some specs for that behaviour.

aliuk2012 commented on spec/features/correspondence_detail_view_spec.rb:
```diff
@@ -35,8 +35,8 @@
     expect(page.internal_deadline).to have_content(internal_gq_deadline)
     expect(page.external_deadline).to have_content(external_gq_deadline)
     expect(page.status).to have_content(gq.state.humanize)
+    expect(page.case_history_list.count).to eq 1
```
Need to check the content

aliuk2012 commented on Gemfile:
```diff
@@ -25,6 +25,7 @@ gem 'uglifier', '>= 1.3.0'
 # gem 'redis', '~> 3.0'
 # Use ActiveModel has_secure_password
 # gem 'bcrypt', '~> 3.1.7'
+gem 'paper_trail'
```
can you place this is alphabetical order?  Could we also add a specific version to it?

aliuk2012 commented on app/views/correspondence/_history.html.slim:
```diff
@@ -0,0 +1,21 @@
+            = happening.who
+        td
+          span#state
+          = happening.state
```
There are quite a few things that need to be added to this markup to make it accessible. I will push my changes to your branch

aliuk2012 replied:
The markup for this file should look something like this:

```
h2
  = 'History'

table
  caption
    span.table-caption.visually-hidden
      | History
  col
  col
  col
  thead
    tr
      th scope='col'
        | When
        span.visually-hidden
          | did the change occur:
      th scope='col'
        | Who
        span.visually-hidden
          | made the change:
      th scope='col'
        span.visually-hidden
          | What
        | State
        span.visually-hidden
          | was it in :
  tbody
    - @correspondence.history.each do | happening |
      tr.case-history-row
        td
          = happening.happened
        td
          = happening.who
        td
          = happening.state
```

aliuk2012 commented on app/views/correspondence/_history.html.slim:
```diff
@@ -0,0 +1,21 @@
+          span#who
+            = happening.who
+        td
+          span#state
```
the ID's for happened,who,state would be invalid if had more than one row. Remember ID's are meant to be unique. 

Is there any particular reason why they were needed? I removed them in my local version and the tests still passed.

aliuk2012 commented on app/views/correspondence/_history.html.slim:
```diff
@@ -0,0 +1,21 @@
+      th State
+  tbody
+    - @correspondence.history.each do | happening |
+      tr.case_history_row
```
I noticed this was being used purely for a testing purpose. If it is could you prefix the class name with 'test-' so we know that this class will be used for test. IMHO I don't think we should  use css selectors only to check content in our tests. Could we check the content some other means....ie like grabbing only part of the page where we kind of expecting the results and then search for the content within that section?

aliuk2012 commented on app/views/correspondence/show.html.slim:
```diff
@@ -30,3 +30,6 @@ section.question.status
     span#status
       = @correspondence.state.humanize
+
+section.question.case_history
```
please change the class to 'case-history'

aliuk2012 commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;hHPU5sw4&#x2F;172-in-the-build-have-tracking-working-so-that-pieces-of-correspondence-have-a-status-which-is-tracked-with-event-history"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> In the build have tracking working, so that pieces of correspondence have a status which is tracked with event history</a>

stephenrichards committed:
CT-1476 BUGFIX Prevent add_message_to_case being called with wrong team

This is a temporary fix to solve the problem of calling an event on a state machine
and passing in an acting team which doesn't have the right to execute that event.

In this case, it is the messages controller, which determines the acting team with:

team = current_user.teams_for_case(@case).first

In the old code, this would have returned one of the teams that was common across
both the case and the user.  If it was the Disclosure team
first, rather than the Disclosure BMT team, then an error occurs because the state
machine rightly says that the action isn't allowed for Disclosure team.

This is a temporary fix to ensure that the messages controller passes in the first
team after they have been ordered in priority 'manager', 'approver', 'responder', which
will work in this case.

But we need to find a more generalised and permanent solution - perhaps asking the
state machine which of the teams has this right.

stephenrichards committed:
Update seed data to include users in teams

This PR adds another team, Communications and Information, and makes Desi Arnaz and Dasha Diss members of that team, as well as Disclosure BMT and Disclosure.

misaka reviewed (approved):
Although, blocked until PR #27 is merged.
